src/arkode/arkode_lsrkstep.c

```diff
@@ -1266,118 +1245,481 @@ int lsrkStep_TakeStepSSPs3(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr
+                          ark_mem->ycur, step_mem->Fe, ark_mem->user_data);
+    step_mem->nfe++;
+    if (retval < 0) { return ARK_RHSFUNC_FAIL; }
+    if (retval > 0) { return RHSFUNC_RECVR; }
+
+#ifdef SUNDIALS_LOGGING_EXTRA_DEBUG
+    SUNLogger_QueueMsg(ARK_LOGGER, SUN_LOGLEVEL_DEBUG,
+                       "ARKODE::lsrkStep_TakeStepSSPs3", "stage RHS",
+                       "F_%i(:) =", j);
+    N_VPrintFile(step_mem->Fe, ARK_LOGGER->debug_fp);
+#endif
+
+    N_VLinearSum(ONE, ark_mem->ycur, ark_mem->h * rat, step_mem->Fe,
+                 ark_mem->ycur);
+    if (!ark_mem->fixedstep)
+    {
+      N_VLinearSum(ONE, ark_mem->tempv1, ark_mem->h / rs, step_mem->Fe,
+                   ark_mem->tempv1);
+    }
+
+    /* apply user-supplied stage postprocessing function (if supplied) */
+    if (ark_mem->ProcessStage != NULL)
+    {
+      retval = ark_mem->ProcessStage(ark_mem->tcur + ((sunrealtype)j - rn - ONE) *
+                                                       rat * ark_mem->h,
+                                     ark_mem->ycur, ark_mem->user_data);
+      if (retval != 0) { return ARK_POSTPROCESS_STAGE_FAIL; }
+    }
+  }
+
+  /* Compute yerr (if step adaptivity enabled) */
+  if (!ark_mem->fixedstep)
+  {
+    N_VLinearSum(ONE, ark_mem->ycur, -ONE, ark_mem->tempv1, ark_mem->tempv1);
+
+    *dsmPtr = N_VWrmsNorm(ark_mem->tempv1, ark_mem->ewt);
+  }
+
+#ifdef SUNDIALS_LOGGING_EXTRA_DEBUG
+  SUNLogger_QueueMsg(ARK_LOGGER, SUN_LOGLEVEL_DEBUG,
+                     "ARKODE::lsrkStep_TakeStepSSPs3", "updated solution",
+                     "ycur(:) =", "");
+  N_VPrintFile(ark_mem->ycur, ARK_LOGGER->debug_fp);
+#endif
+
+#if SUNDIALS_LOGGING_LEVEL >= SUNDIALS_LOGGING_DEBUG
+  SUNLogger_QueueMsg(ARK_LOGGER, SUN_LOGLEVEL_DEBUG,
+                     "ARKODE::lsrkStep_TakeStepSSPs3", "error-test",
+                     "step = %li, h = %" RSYM ", dsm = %" RSYM, ark_mem->nst,
+                     ark_mem->h, *dsmPtr);
+#endif
+
+  return ARK_SUCCESS;
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_TakeStepSSP43:
+
+  This routine serves the primary purpose of the LSRKStepSSP43 module:
+  it performs a single SSP43 step (with embedding).
+
+  The output variable dsmPtr should contain estimate of the
+  weighted local error if an embedding is present; otherwise it
+  should be 0.
+
+  The input/output variable nflagPtr is used to gauge convergence
+  of any algebraic solvers within the step.  As this routine
+  involves no algebraic solve, it is set to 0 (success).
+
+  The return value from this routine is:
+            0 => step completed successfully
+           >0 => step encountered recoverable failure;
+                 reduce step and retry (if possible)
+           <0 => step encountered unrecoverable failure
+  ---------------------------------------------------------------*/
+int lsrkStep_TakeStepSSP43(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
```

Comment:
We used to require at least 9 stages before, and extended this with the case s=4. This case was differing significantly in the structure, so I wanted to keep it separately instead of having zillions of ifs/elses.

Reply:
This comment was before I realized `SSP43` is a special case of `SSPs3`. This does not need be documented in the user guide but some comments should be added here and before the SSPs3 take step function to note this take step function is used when there are only four stages.

Reply:
Comments are there now.